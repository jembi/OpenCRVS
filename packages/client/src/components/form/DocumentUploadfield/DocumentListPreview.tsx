/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import * as React from 'react'
import styled from 'styled-components'
import { IFileValue, IAttachmentValue } from '@client/forms'
import { Spinner } from '@opencrvs/components/lib/Spinner'
import { ISelectOption } from '@opencrvs/components/lib/Select'
import { Link } from '@opencrvs/components/lib/Link/Link'
import { Icon } from '@opencrvs/components/lib/Icon/Icon'
import { Button } from '@opencrvs/components/lib/Button/Button'

const Wrapper = styled.div`
  margin: 8px 0px;
  max-width: 100%;
`

const Container = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 4px;
  border-bottom: 1px solid ${({ theme }) => theme.colors.grey200};
  height: 40px;
  padding: 0px 8px;
`

const SpinnerContainer = styled(Spinner)`
  margin-right: 6px;
`

const DeleteContainer = styled(Button)`
  min-width: 32px;
`

const Label = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-start;
  gap: 8px;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
  svg {
    flex-shrink: 0;
  }
`

type Props = {
  id?: string
  documents?: IFileValue[] | null
  processingDocuments?: Array<{ label: string }>
  attachment?: IAttachmentValue
  label?: string
  onSelect: (document: IFileValue | IAttachmentValue) => void
  dropdownOptions?: ISelectOption[]
  onDelete?: (image: IFileValue | IAttachmentValue) => void
  inReviewSection?: boolean
}

export const DocumentListPreview = ({
  id,
  documents,
  processingDocuments,
  attachment,
  label,
  onSelect,
  dropdownOptions,
  onDelete,
  inReviewSection
}: Props) => {
  const getFormattedLabelForDocType = (docType: string) => {
    const matchingOptionForDocType =
      dropdownOptions &&
      dropdownOptions.find((option) => option.value === docType)
    return matchingOptionForDocType && matchingOptionForDocType.label
  }

  return (
    <Wrapper id={`preview-list-${id}`}>
      {documents &&
        documents.map((document: IFileValue, key: number) => (
          <Container key={`preview_${key}`}>
            <Label>
              <Icon color="currentColor" name="Paperclip" size="medium" />
              <Link
                id={`document_${(document.optionValues[1] as string).replace(
                  /\s/g,
                  ''
                )}_link`}
                key={key}
                onClick={(_) => onSelect(document)}
              >
                <span>
                  {(inReviewSection &&
                    dropdownOptions &&
                    dropdownOptions[key]?.label) ||
                    getFormattedLabelForDocType(
                      document.optionValues[1] as string
                    ) ||
                    (document.optionValues[1] as string)}
                </span>
              </Link>
            </Label>
            {onDelete && (
              <DeleteContainer
                id="preview_delete"
                type="icon"
                size="small"
                aria-label="Delete attachment"
                onClick={() => onDelete(document)}
              >
                <Icon color="red" name="TrashSimple" size="small" />
              </DeleteContainer>
            )}
          </Container>
        ))}
      {processingDocuments &&
        processingDocuments.map(({ label }) => (
          <Container key={label}>
            <Label>
              <Icon color="disabled" name="Paperclip" size="medium" />
              <Link disabled={true} key={label}>
                <span>{getFormattedLabelForDocType(label) || label}</span>
              </Link>
            </Label>
            <SpinnerContainer size={20} id={`document_${label}_processing`} />
          </Container>
        ))}
      {attachment && attachment.data && label && (
        <Container>
          <Label>
            <Icon color="grey600" name="Paperclip" size="medium" />
            <Link onClick={(_) => onSelect(attachment)}>
              <span>{getFormattedLabelForDocType(label) || label}</span>
            </Link>
          </Label>
          <DeleteContainer
            id="preview_delete"
            type="icon"
            size="small"
            aria-label="Delete attachment"
            onClick={() => onDelete && onDelete(attachment)}
          >
            <Icon color="red" name="TrashSimple" size="small" />
          </DeleteContainer>
        </Container>
      )}
    </Wrapper>
  )
}
