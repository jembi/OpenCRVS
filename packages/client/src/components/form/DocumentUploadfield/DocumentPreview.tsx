/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import * as React from 'react'
import styled from 'styled-components'
import { IFileValue, IAttachmentValue } from '@client/forms'
import { AppBar } from '@opencrvs/components/lib/AppBar'
import { Stack } from '@opencrvs/components/lib/Stack'
import { Button } from '@opencrvs/components/lib/Button'
import { Icon } from '@opencrvs/components/lib/Icon'
import { DividerVertical } from '@opencrvs/components/lib/Divider'
import PanControls from '@opencrvs/components/lib/DocumentViewer/components/PanControls'
import PanViewer from '@opencrvs/components/lib/DocumentViewer/components/PanViewer'
import { useState } from 'react'

const ViewerWrapper = styled.div`
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 4;
  width: 100%;
  height: 100%;
  background: ${({ theme }) => theme.colors.white};
`

const ViewerContainer = styled.div`
  position: relative;
  width: 100%;
  height: 100%;
  overflow: hidden;
  display: flex;
  align-items: center;
  & img {
    max-height: 80vh;
    max-width: 80vw;
    width: auto;
  }
`

interface IProps {
  previewImage: IFileValue | IAttachmentValue
  disableDelete?: boolean
  title?: string
  goBack: () => void
  onDelete: (image: IFileValue | IAttachmentValue) => void
  id?: string
}

export const DocumentPreview = ({
  previewImage,
  title,
  goBack,
  onDelete,
  disableDelete,
  id
}: IProps) => {
  const [zoom, setZoom] = useState(1)
  const [rotation, setRotation] = useState(0)

  const zoomIn = () => setZoom((prevState) => prevState + 0.2)
  const zoomOut = () =>
    setZoom((prevState) => (prevState >= 1 ? prevState - 0.2 : prevState))
  const rotateLeft = () => setRotation((prevState) => (prevState - 90) % 360)

  return (
    <ViewerWrapper id={id ?? 'preview_image_field'}>
      <AppBar
        desktopLeft={<Icon name="Paperclip" size="large" />}
        desktopTitle={title}
        desktopRight={
          <Stack gap={8}>
            <PanControls
              zoomIn={zoomIn}
              zoomOut={zoomOut}
              rotateLeft={rotateLeft}
            />
            {!disableDelete && (
              <>
                <DividerVertical />
                <Button
                  id="preview_delete"
                  type="icon"
                  onClick={() => onDelete(previewImage)}
                >
                  <Icon name="TrashSimple" color="red" />
                </Button>
              </>
            )}
            <DividerVertical />
            <Button
              id="preview_close"
              aria-label="Go close"
              size="medium"
              type="icon"
              onClick={goBack}
            >
              <Icon name="X" size="medium" />
            </Button>
          </Stack>
        }
        mobileLeft={<Icon name="Paperclip" size="large" />}
        mobileTitle={title}
        mobileRight={
          <Stack gap={8}>
            <PanControls
              zoomIn={zoomIn}
              zoomOut={zoomOut}
              rotateLeft={rotateLeft}
            />
            {!disableDelete && (
              <Button
                id="preview_delete"
                type="icon"
                onClick={() => onDelete(previewImage)}
              >
                <Icon name="TrashSimple" color="red" />
              </Button>
            )}
            <Button
              id="preview_close"
              aria-label="Go back"
              size="medium"
              type="icon"
              onClick={goBack}
            >
              <Icon name="X" size="medium" />
            </Button>
          </Stack>
        }
      />

      <ViewerContainer>
        {previewImage.data && (
          <PanViewer
            key={Math.random()}
            id="document_image"
            image={previewImage.data}
            zoom={zoom}
            rotation={rotation}
          />
        )}
      </ViewerContainer>
    </ViewerWrapper>
  )
}
