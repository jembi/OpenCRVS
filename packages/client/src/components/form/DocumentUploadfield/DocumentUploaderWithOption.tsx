/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { ISelectOption, Select } from '@opencrvs/components/lib/Select'
import { ImageUploader } from '@opencrvs/components/lib/ImageUploader'
import { ErrorText } from '@opencrvs/components/lib/ErrorText'
import { DocumentPreview } from '@client/components/form/DocumentUploadfield/DocumentPreview'
import { IFileValue, IFormFieldValue, IAttachmentValue } from '@client/forms'
import { ALLOWED_IMAGE_TYPE, EMPTY_STRING } from '@client/utils/constants'
import * as React from 'react'
import {
  WrappedComponentProps as IntlShapeProps,
  injectIntl,
  MessageDescriptor
} from 'react-intl'
import styled from 'styled-components'
import { DocumentListPreview } from './DocumentListPreview'
import { remove, clone } from 'lodash'
import { buttonMessages, formMessages } from '@client/i18n/messages'
import { messages } from '@client/i18n/messages/views/imageUpload'
import imageCompression from 'browser-image-compression'

const options = {
  maxSizeMB: 0.4,
  maxWidthOrHeight: 1920,
  useWebWorker: true
}

const UploaderWrapper = styled.div`
  margin-bottom: 28px;
`

const Label = styled.label`
  position: relative;
  color: ${({ theme }) => theme.colors.copy};
  ${({ theme }) => theme.fonts.reg18};
`
const Flex = styled.div<{ splitView?: boolean }>`
  display: flex;
  flex-wrap: nowrap;
  gap: 8px;
  margin-bottom: ${({ splitView }) => {
    return splitView ? '10px' : '0px'
  }};
`
export const ErrorMessage = styled.div`
  margin-bottom: 16px;
`

type IFullProps = {
  name: string
  label: string
  placeholder?: string
  extraValue: IFormFieldValue
  options: ISelectOption[]
  splitView?: boolean
  files: IFileValue[]
  hideOnEmptyOption?: boolean
  onComplete: (files: IFileValue[]) => void
  touched?: boolean
  onUploadingStateChanged?: (isUploading: boolean) => void
  requiredErrorMessage?: MessageDescriptor
} & IntlShapeProps

type DocumentFields = {
  documentType: string
  documentData: string
}

type IState = {
  errorMessage: string
  fields: DocumentFields
  previewImage: IFileValue | null
  filesBeingProcessed: Array<{ label: string }>
  dropDownOptions: ISelectOption[]
}

export const getBase64String = (file: File) => {
  return new Promise<string | ArrayBuffer>((resolve, reject) => {
    const reader = new FileReader()
    reader.readAsDataURL(file)
    reader.onload = () => {
      if (reader.result) {
        return resolve(reader.result)
      }
    }
    reader.onerror = (error) => reject(error)
  })
}

class DocumentUploaderWithOptionComp extends React.Component<
  IFullProps,
  IState
> {
  constructor(props: IFullProps) {
    super(props)
    this.state = {
      errorMessage: EMPTY_STRING,
      previewImage: null,
      dropDownOptions: this.initializeDropDownOption(),
      filesBeingProcessed: [],
      fields: {
        documentType: EMPTY_STRING,
        documentData: EMPTY_STRING
      }
    }
  }

  initializeDropDownOption = (): ISelectOption[] => {
    const options = clone(this.props.options)
    this.props.files &&
      this.props.files.forEach((element: IFileValue) => {
        remove(
          options,
          (option: ISelectOption) => option.value === element.optionValues[1]
        )
      })

    return options
  }

  onChange = (documentType: string) => {
    const currentState = this.state
    currentState.fields.documentType = documentType
    this.setState(currentState)
  }

  isValid = (): boolean => {
    const isValid = !!this.state.fields.documentType

    this.setState({
      errorMessage: isValid
        ? EMPTY_STRING
        : this.props.intl.formatMessage(messages.documentTypeRequired)
    })

    return isValid
  }

  processImage = async (uploadedImage: File) => {
    if (!ALLOWED_IMAGE_TYPE.includes(uploadedImage.type)) {
      this.setState({
        errorMessage: this.props.intl.formatMessage(messages.uploadError)
      })
      throw new Error('File type not supported')
    }

    if (uploadedImage.size > 5242880) {
      this.setState({
        errorMessage: this.props.intl.formatMessage(messages.overSized)
      })
      throw new Error(this.props.intl.formatMessage(messages.overSized))
    }

    const resized =
      uploadedImage.size > 512000 &&
      (await imageCompression(uploadedImage, options))

    const fileAsBase64 = await getBase64String(resized || uploadedImage)

    return fileAsBase64.toString()
  }

  handleFileChange = async (uploadedImage: File) => {
    if (!uploadedImage) {
      return
    }

    let fileAsBase64: string
    const optionValues: [IFormFieldValue, string] = [
      this.props.extraValue,
      this.state.fields.documentType
    ]

    this.setState((state) => ({
      filesBeingProcessed: [
        ...state.filesBeingProcessed,
        {
          label: optionValues[1]
        }
      ]
    }))

    this.props.onUploadingStateChanged &&
      this.props.onUploadingStateChanged(true)

    const minimumProcessingTime = new Promise<void>((resolve) =>
      setTimeout(resolve, 2000)
    )

    try {
      // Start processing
      ;[fileAsBase64] = await Promise.all([
        this.processImage(uploadedImage),
        minimumProcessingTime
      ])
    } catch (error) {
      this.props.onUploadingStateChanged &&
        this.props.onUploadingStateChanged(false)

      this.setState({
        errorMessage:
          this.state.errorMessage ||
          this.props.intl.formatMessage(messages.uploadError),
        // Remove from processing files
        filesBeingProcessed: this.state.filesBeingProcessed.filter(
          ({ label }) => label !== optionValues[1]
        )
      })
      return
    }

    const tempOptions = this.state.dropDownOptions

    remove(
      tempOptions,
      (option: ISelectOption) => option.value === this.state.fields.documentType
    )

    const newDocument: IFileValue = {
      optionValues,
      type: uploadedImage.type,
      data: fileAsBase64.toString(),
      fileSize: uploadedImage.size
    }

    this.props.onComplete([...this.props.files, newDocument])
    this.props.onUploadingStateChanged &&
      this.props.onUploadingStateChanged(false)

    this.setState((prevState) => {
      return {
        ...prevState,
        errorMessage: EMPTY_STRING,
        fields: {
          documentType: EMPTY_STRING,
          documentData: EMPTY_STRING
        },
        dropDownOptions: tempOptions,
        // Remove from processing files
        filesBeingProcessed: this.state.filesBeingProcessed.filter(
          ({ label }) => label !== optionValues[1]
        )
      }
    })
  }

  onDelete = (image: IFileValue | IAttachmentValue) => {
    const previewImage = image as IFileValue
    const addableOption = this.props.options.find(
      (item: ISelectOption) => item.value === previewImage.optionValues[1]
    ) as ISelectOption
    const dropDownOptions = this.state.dropDownOptions.concat(addableOption)
    this.setState(() => ({ dropDownOptions }))
    this.props.onComplete(
      this.props.files.filter((file) => file !== previewImage)
    )
    this.closePreviewSection()
  }

  closePreviewSection = () => {
    this.setState({ previewImage: null })
  }

  selectForPreview = (previewImage: IFileValue | IAttachmentValue) => {
    this.setState({ previewImage: previewImage as IFileValue })
  }

  getFormattedLabelForDocType = (docType: string) => {
    const matchingOptionForDocType =
      this.props.options &&
      this.props.options.find((option) => option.value === docType)
    return matchingOptionForDocType && matchingOptionForDocType.label
  }

  renderDocumentUploaderWithDocumentTypeBlock = () => {
    const { name, intl, placeholder } = this.props
    return this.props.splitView ? (
      this.state.dropDownOptions.map((opt, idx) => (
        <Flex splitView key={idx}>
          <Select
            id={`${name}${idx}`}
            placeholder={placeholder}
            options={[opt]}
            value={opt.value}
            onChange={this.onChange}
          />

          <ImageUploader
            id={`upload_document${idx}`}
            title={intl.formatMessage(formMessages.addFile)}
            onClick={(e) => {
              this.onChange(opt.value)
              return !this.isValid() && e.preventDefault()
            }}
            handleFileChange={this.handleFileChange}
            disabled={this.state.filesBeingProcessed.length > 0}
          />
        </Flex>
      ))
    ) : (
      <Flex>
        <Select
          id={name}
          placeholder={placeholder}
          options={this.state.dropDownOptions}
          value={this.state.fields.documentType}
          onChange={this.onChange}
          isDisabled={this.state.filesBeingProcessed.length > 0}
        />

        <ImageUploader
          id="upload_document"
          title={intl.formatMessage(formMessages.addFile)}
          onClick={(e) => !this.isValid() && e.preventDefault()}
          handleFileChange={this.handleFileChange}
          disabled={this.state.filesBeingProcessed.length > 0}
        />
      </Flex>
    )
  }

  render() {
    const { label, intl, requiredErrorMessage } = this.props

    return (
      <UploaderWrapper>
        <ErrorMessage id="upload-error">
          {this.state.errorMessage && (
            <ErrorText>
              {(requiredErrorMessage &&
                intl.formatMessage(requiredErrorMessage)) ||
                this.state.errorMessage}
            </ErrorText>
          )}
        </ErrorMessage>

        <Label>{label}</Label>
        <DocumentListPreview
          processingDocuments={this.state.filesBeingProcessed}
          documents={this.props.files}
          onSelect={this.selectForPreview}
          dropdownOptions={this.props.options}
          onDelete={this.onDelete}
        />
        {this.props.hideOnEmptyOption && this.state.dropDownOptions.length === 0
          ? null
          : this.renderDocumentUploaderWithDocumentTypeBlock()}

        {this.state.previewImage && (
          <DocumentPreview
            previewImage={this.state.previewImage}
            title={this.getFormattedLabelForDocType(
              this.state.previewImage.optionValues[1] as string
            )}
            goBack={this.closePreviewSection}
            onDelete={this.onDelete}
          />
        )}
      </UploaderWrapper>
    )
  }
}

export const DocumentUploaderWithOption = injectIntl<'intl', IFullProps>(
  DocumentUploaderWithOptionComp
)
