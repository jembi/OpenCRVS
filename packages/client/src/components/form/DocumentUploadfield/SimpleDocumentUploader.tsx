/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { ImageUploader } from '@opencrvs/components/lib/ImageUploader'
import { ErrorText } from '@opencrvs/components/lib/ErrorText'
import { DocumentPreview } from '@client/components/form/DocumentUploadfield/DocumentPreview'
import { IFormFieldValue, IAttachmentValue } from '@client/forms'
import * as React from 'react'
import {
  WrappedComponentProps as IntlShapeProps,
  injectIntl,
  MessageDescriptor
} from 'react-intl'
import styled from 'styled-components'
import { DocumentListPreview } from './DocumentListPreview'
import { buttonMessages, formMessages as messages } from '@client/i18n/messages'
import { getBase64String, ErrorMessage } from './DocumentUploaderWithOption'

const DocumentUploader = styled(ImageUploader)`
  color: ${({ theme }) => theme.colors.primary};
  background: ${({ theme }) => theme.colors.white};
  border: ${({ theme }) => `2px solid ${theme.colors.primary}`};
  border-radius: 4px;
  ${({ theme }) => theme.fonts.bold14};
  height: 40px;
  text-transform: initial;

  @media (max-width: ${({ theme }) => theme.grid.breakpoints.md}px) {
    margin-left: 0px;
    margin-top: 10px;
  }
`

const FieldDescription = styled.div`
  margin-top: 0px;
  margin-bottom: 6px;
`

type IFullProps = {
  name: string
  label: string
  files?: IAttachmentValue
  description?: string
  allowedDocType?: string[]
  error?: string
  disableDeleteInPreview?: boolean
  onComplete: (files: IAttachmentValue | {}) => void
  touched?: boolean
  onUploadingStateChanged?: (isUploading: boolean) => void
  requiredErrorMessage?: MessageDescriptor
  previewTransformer?: (files: IAttachmentValue) => IAttachmentValue
} & IntlShapeProps

type IState = {
  error: string
  previewImage: IAttachmentValue | null
  filesBeingUploaded: Array<{ label: string }>
}
class SimpleDocumentUploaderComponent extends React.Component<
  IFullProps,
  IState
> {
  constructor(props: IFullProps) {
    super(props)
    this.state = {
      error: '',
      previewImage: null,
      filesBeingUploaded: []
    }
  }

  handleFileChange = async (uploadedImage: File) => {
    if (!uploadedImage) {
      return
    }
    const allowedDocType = this.props.allowedDocType

    this.setState(() => ({
      filesBeingUploaded: [
        ...this.state.filesBeingUploaded,
        {
          label: uploadedImage.name
        }
      ]
    }))

    this.props.onUploadingStateChanged &&
      this.props.onUploadingStateChanged(true)

    if (
      allowedDocType &&
      allowedDocType.length > 0 &&
      !allowedDocType.includes(uploadedImage.type)
    ) {
      this.props.onUploadingStateChanged &&
        this.props.onUploadingStateChanged(false)
      this.setState({
        filesBeingUploaded: []
      })
      this.setState({
        error: this.props.intl.formatMessage(messages.fileUploadError, {
          type: allowedDocType
            .map((docTypeStr) => docTypeStr.split('/').pop())
            .join(', ')
        })
      })
    } else {
      this.props.onUploadingStateChanged &&
        this.props.onUploadingStateChanged(false)
      this.props.onComplete({
        name: uploadedImage.name,
        type: uploadedImage.type,
        data: await getBase64String(uploadedImage)
      })
      this.setState({
        error: ''
      })
      this.setState({
        filesBeingUploaded: []
      })
    }
  }

  selectForPreview = (previewImage: IFormFieldValue) => {
    if (this.props.previewTransformer) {
      return this.setState({
        previewImage: this.props.previewTransformer(
          previewImage as IAttachmentValue
        )
      })
    }
    this.setState({ previewImage: previewImage as IAttachmentValue })
  }

  closePreviewSection = () => {
    this.setState({ previewImage: null })
  }

  onDelete = (image: IFormFieldValue) => {
    this.props.onComplete('')
    this.closePreviewSection()
  }

  render() {
    const {
      label,
      intl,
      files,
      description,
      error,
      disableDeleteInPreview,
      requiredErrorMessage,
      touched
    } = this.props
    const errorMessage =
      (requiredErrorMessage && intl.formatMessage(requiredErrorMessage)) ||
      this.state.error ||
      error ||
      ''

    return (
      <>
        {description && <FieldDescription>{description}</FieldDescription>}
        <ErrorMessage>
          {errorMessage && (touched || this.state.error) && (
            <ErrorText ignoreMediaQuery id="field-error">
              {errorMessage}
            </ErrorText>
          )}
        </ErrorMessage>
        <DocumentListPreview
          attachment={files}
          onSelect={this.selectForPreview}
          label={label}
          onDelete={this.onDelete}
          processingDocuments={this.state.filesBeingUploaded}
        />
        {this.state.previewImage && (
          <DocumentPreview
            previewImage={this.state.previewImage}
            disableDelete={disableDeleteInPreview}
            title={intl.formatMessage(buttonMessages.preview)}
            goBack={this.closePreviewSection}
            onDelete={this.onDelete}
          />
        )}
        {(!files || !files.data) && (
          <DocumentUploader
            id="upload_document"
            title={intl.formatMessage(messages.uploadFile)}
            handleFileChange={this.handleFileChange}
          />
        )}
      </>
    )
  }
}

export const SimpleDocumentUploader = injectIntl<'intl', IFullProps>(
  SimpleDocumentUploaderComponent
)
