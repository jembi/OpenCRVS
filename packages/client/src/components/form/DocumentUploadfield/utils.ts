import { getBase64String, ErrorMessage } from './DocumentUploaderWithOption'
import Jimp from 'jimp'
import { buttonMessages, formMessages as messages } from '@client/i18n/messages'
import { IFormFieldValue, IAttachmentValue } from '@client/forms'

export const handleFileChange = (thisObj: any, uploadedImage: File) => {
  if (!uploadedImage) {
    return
  }
  const allowedDocType = thisObj.props.allowedDocType

  thisObj.setState(() => ({
    filesBeingUploaded: [
      ...thisObj.state.filesBeingUploaded,
      {
        label: uploadedImage.name
      }
    ]
  }))

  thisObj.props.onUploadingStateChanged &&
    thisObj.props.onUploadingStateChanged(true)

  getBase64String(uploadedImage).then((data) => {
    let base64String = data as string
    base64String = base64String.split('base64,')[1]
    Jimp.read(new Buffer(base64String, 'base64'))
      .then((buffer) => {
        if (
          allowedDocType &&
          allowedDocType.length > 0 &&
          !allowedDocType.includes(buffer.getMIME())
        ) {
          throw new Error('File type not supported')
        }
        return data as string
      })
      .then((buffer) => {
        thisObj.props.onUploadingStateChanged &&
          thisObj.props.onUploadingStateChanged(false)
        thisObj.props.onComplete({
          name: uploadedImage.name,
          type: uploadedImage.type,
          data: buffer
        })
        thisObj.setState({
          error: ''
        })
        thisObj.setState({
          filesBeingUploaded: []
        })
      })
      .catch(() => {
        thisObj.props.onUploadingStateChanged &&
          thisObj.props.onUploadingStateChanged(false)
        thisObj.setState({
          filesBeingUploaded: []
        })
        allowedDocType &&
          allowedDocType.length > 0 &&
          thisObj.setState({
            error: thisObj.props.intl.formatMessage(messages.fileUploadError, {
              type: allowedDocType
                .map((docTypeStr: string) => docTypeStr.split('/').pop())
                .join(', ')
            })
          })
      })
  })
}

export const selectForPreview = (
  thisObj: any,
  previewImage: IFormFieldValue
) => {
  thisObj.setState({ previewImage: previewImage as IAttachmentValue })
}

export const closePreviewSection = (thisObj: any) => {
  thisObj.setState({ previewImage: null })
}

export const onDelete = (thisObj: any, image: IFormFieldValue) => {
  thisObj.props.onComplete('')
  thisObj.closePreviewSection()
}
