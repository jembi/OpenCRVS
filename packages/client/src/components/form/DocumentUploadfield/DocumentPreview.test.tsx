/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import * as React from 'react'
import { createShallowRenderedComponent } from '@client/tests/util'
import { DocumentPreview } from '@client/components/form/DocumentUploadfield/DocumentPreview'
import { vi } from 'vitest'

describe('image upload component', () => {
  const mock = vi.fn()
  const testComponent = createShallowRenderedComponent(
    <DocumentPreview
      previewImage={{
        optionValues: [],
        type: 'image/png',
        data: 'dummy base64 data'
      }}
      goBack={mock}
      onDelete={mock}
    />
  )

  it('renders without crashing', () => {
    expect(testComponent).toMatchSnapshot()
  })
})
