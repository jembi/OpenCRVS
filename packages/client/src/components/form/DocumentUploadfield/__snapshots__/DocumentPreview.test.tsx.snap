// Vitest Snapshot v1

exports[`image upload component > renders without crashing 1`] = `ShallowWrapper {}`;

exports[`image upload component renders without crashing 1`] = `ShallowWrapper {}`;
